reader_concurrency_semaphore.cc

```diff
@@ -87,7 +87,8 @@ class reader_permit::impl
     bool _marked_as_used = false;
     uint64_t _blocked_branches = 0;
     bool _marked_as_blocked = false;
-    db::timeout_clock::time_point _timeout;
+    abort_source _as;
+    timer<db::timeout_clock> _ttl_timer;
```

Comment:
I'm not thrilled about abort_source here. I keep harassing @bhalevy how abort_source is an administrative object, and now it's used here per-request.

Reply:
The logic is that if abort_source is an administrative object, we can make it fatter and easier to use without regretting the effect on the data plane.

Reply:
I don't understand the administrative vs. data plane distinction.
I can replace it with just a bool and open coding select parts of abort_source (check_aborted()).

Reply:
administrative = constructor/destructor/abort called on rare path like service up/down (checking abort can happen per request)
data plane = constructor/destructor/abort/check called on every request

If we allow data plane abort source, it means we must be very careful about its memory and code footprint.
If we only allow administrative abort_source, we can go wild with improving its ease of use.

Reply:
Replaced with `std::exception_ptr` in v2.